presto-main/src/main/java/com/facebook/presto/operator/scalar/ReflectionParametricScalar.java

```diff
@@ -0,0 +1,625 @@
+        boolean hidden;
+        boolean deterministic;
+
+        if (scalarAnnotation != null) {
+            name = scalarAnnotation.value();
+            hidden = scalarAnnotation.hidden();
+            deterministic = scalarAnnotation.deterministic();
+        }
+        else {
+            name = mangleOperatorName(operatorAnnotation.value());
+            hidden = true;
+            deterministic = true;
+        }
+
+        ImmutableMap.Builder<Signature, Implementation> exactImplementations = ImmutableMap.builder();
+        ImmutableList.Builder<Implementation> specializedImplementations = ImmutableList.builder();
+        ImmutableList.Builder<Implementation> implementations = ImmutableList.builder();
+        Signature signature = null;
+
+        for (Method method : findPublicMethodsWithAnnotation(clazz, SqlType.class)) {
+            Implementation implementation = parseImplementation(name, method, clazz.getConstructors());
+            if (implementation.getSignature().getTypeParameters().isEmpty()) {
+                exactImplementations.put(implementation.getSignature(), implementation);
+                continue;
+            }
+            if (signature == null) {
+                signature = implementation.getSignature();
+            }
+            else {
+                checkArgument(implementation.getSignature().equals(signature), "Implementations with type parameters must all have matching signatures. %s does not match %s", implementation.getSignature(), signature);
+            }
+            if (implementation.hasSpecializedTypeParameters()) {
+                specializedImplementations.add(implementation);
+            }
+            else {
+                implementations.add(implementation);
+            }
+        }
+
+        requireNonNull(signature, format("No implementations found for %s", name));
+
+        return new ReflectionParametricScalar(
+                signature,
+                description,
+                hidden,
+                exactImplementations.build(),
+                specializedImplementations.build(),
+                implementations.build(),
+                deterministic);
+    }
+
+    private static List<Method> findPublicMethodsWithAnnotation(Class<?> clazz, Class<?> annotationClass)
+    {
+        ImmutableList.Builder<Method> methods = ImmutableList.builder();
+        for (Method method : clazz.getMethods()) {
+            for (Annotation annotation : method.getAnnotations()) {
+                if (annotationClass.isInstance(annotation)) {
+                    checkArgument(Modifier.isPublic(method.getModifiers()), "%s annotated with %s must be public", method.getName(), annotationClass.getSimpleName());
+                    methods.add(method);
+                }
+            }
+        }
+        return methods.build();
+    }
+
+    private static Implementation parseImplementation(String name, Method method, Constructor<?>[] constructors)
```

Comment:
Do you think this function is too long? I considered breaking it up, but because of all the different lists that have to be built it would be difficult to do without having the helper methods use out variables